### PR TITLE
Add matcher config in configuration file

### PR DESCRIPTION
**Pull request by mbj**

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -48,6 +48,42 @@ When `fail_fast` is enabled, mutant will stop as soon as it encounters an alive
...
+matcher:
+  # Subject expressions to find subjects for mutation testing.
+  # Multiple entries are allowed and matches from each expression
+  # are unioned.
+  #
+  # Subject expressions can also be specified on the command line. Example:
+  # `bundle exec mutant run YourSubject`
+  #
+  # Note that expressions from the command line replace the subjects
+  # configured # in the config file!
```
Comment (dgollahon) [inline docs/configuration.md:65]: Looks like an errant `#` here.

Comment (mbj) [reply docs/configuration.md]: fix & fixed.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -43,7 +43,7 @@ jobs:
       with:
         ruby-version: ${{ matrix.ruby }}
         bundler-cache: true
-    - run: ./mutant.sh --since HEAD~1 -- 'Mutant*'
+    - run: bundle exec mutant run --zombie --since HEAD~1
```
Comment (dgollahon) [inline .github/workflows/ci.yml:46]: Should `mutant.sh` be deleted?

File: lib/mutant/config.rb [hunk 3]
```diff
@@ -159,13 +112,14 @@ def self.env
         Transform::Exception.new(YAML::SyntaxError, YAML.method(:safe_load)),
         Transform::Hash.new(
           optional: [
-            Transform::Hash::Key.new('coverage_criteria', CoverageCriteria::TRANSFORM),
+            Transform::Hash::Key.new('coverage_criteria', ->(value) { CoverageCriteria::TRANSFORM.call(value) }),
```
Comment (mbj) [inline lib/mutant/config.rb:115]: I "really" like this require sequence:

```
require 'foo'
require 'foo/bar'
```

But we have to access the constant from `foo/bar` here, hence as we are in class scope and the require for `foo/bar' did not happen yet we have to use a lambda indirection.

The alternative would be:

```
require 'foo/bar'
require 'foo'
```
which sucks. For the moment I eat it. Would love of Ruby had a module system ;).

File: docs/mutant-minitest.md [hunk 4]
```diff
@@ -64,7 +64,7 @@ This prints a report like:
 
 ```sh
 Mutant environment:
-Matcher:         #<Mutant::Matcher::Config match_expressions: [AUOM*]>
+Matcher:         #<Mutant::Matcher::Config subjects: [AUOM*]>
```
Comment (dgollahon) [inline docs/mutant-minitest.md:67]: This is a good change I think--it reads more clearly.

Comment (mbj) [inline docs/mutant-minitest.md:67]: I think this is much nicer to read. Its clear the domain of this method is matching, else the class would not be named `Mutant::Matcher`.

So the `match_` prefix was redundant always.

Just naming the key `expressions` would IMO also suck as it than would have a sibling `ignore_expressions` which does not clearly indicate what the difference between both would be.

Hence replacing this with `subjects` makes it very clear. And `subject_expressions` is kind of redundant as we can only match expressions.

File: lib/mutant.rb [hunk 5]
```diff
@@ -41,6 +41,7 @@ module Mutant
   SCOPE_OPERATOR = '::'
 end # Mutant
 
+require 'mutant/transform'
```
Comment (mbj) [inline lib/mutant.rb:44]: The fact this has to move up the require chain to toplevel, indicates it likely belongs to a separate library.

Once I make nicer "input error location rendering" this is worthy of a separate lib.

File: lib/mutant/config.rb (hunk 3 above)
Comment (dgollahon) [reply lib/mutant/config.rb:115]: > Would love of Ruby had a module system ;)

yeahh... I feel that one!

File: lib/mutant/config.rb [hunk 6]
```diff
@@ -159,13 +112,14 @@ def self.env
...
             Transform::Hash::Key.new('fail_fast',         Transform::BOOLEAN),
             Transform::Hash::Key.new('includes',          Transform::STRING_ARRAY),
             Transform::Hash::Key.new('integration',       Transform::STRING),
             Transform::Hash::Key.new('jobs',              Transform::INTEGER),
             Transform::Hash::Key.new('mutation_timeout',  Transform::FLOAT),
-            Transform::Hash::Key.new('requires',          Transform::STRING_ARRAY)
+            Transform::Hash::Key.new('requires',          Transform::STRING_ARRAY),
+            Transform::Hash::Key.new('matcher',           Matcher::Config::LOADER)
```
Comment (mbj) [inline lib/mutant/config.rb:122]: This one, paired with adding the constant `Mutant::Config::LOADER` is the only real change to make config file support for matchers.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (mbj) [reply .github/workflows/ci.yml:46]: yes, good catch. Happy to se it go.

I fought myself to add `zombie: true` to the config file loader, and decided against it.

Nobody who is not willing to read mutants source should ever use `--zombie`.

Comment (dgollahon) [reply .github/workflows/ci.yml:46]: > I fought myself to add zombie: true to the config file loader, and decided against it.

That seems like the right call.

**Review by dgollahon — DISMISSED** [review]
Woo! I'm excited about this one 🎉 LGTM assuming CI passes.

**Comment by mbj** [comment]
> Woo! I'm excited about this one  LGTM assuming CI passes.

it should, I repushed a lot so have to manage the queue. Merging once done.

**Pull request merged**